File: frontend/src/api/documents.ts
import { apiClient, API_BASE_URL } from "./client";

export interface DocumentSummary {
  id: string;
  title: string;
  original_filename: string;
  file_type: string;
  file_size_bytes: number;
  ocr_status: "PENDING" | "PROCESSING" | "DONE" | "APPROVED" | "REJECTED" | "FAILED" | "SKIPPED";
  minio_object_key: string;
  created_at: string;
}

export interface DocumentListItem extends DocumentSummary {
  page_count?: number;
  category_id?: string;
  uploaded_by: string;
  is_deleted: boolean;
  updated_at: string;
  uploader_name?: string;
  uploader_mssv?: string;
  category_name?: string;
  confidence_score?: number;
}

export interface DocumentListResponse {
  items: DocumentListItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface OCRResult {
  id: string;
  document_id: string;
  is_latest: boolean;
  raw_text?: string;
  corrected_text?: string;
  confidence_score?: number;
  ocr_engine: string;
  is_corrected: boolean;
  corrected_by?: string;
  corrected_at?: string;
  processing_time_ms?: number;
  created_at: string;
}

export interface ProcessingJob {
  id: string;
  status: "PENDING" | "RUNNING" | "SUCCESS" | "FAILED";
  error_message?: string;
  retry_count: number;
  started_at?: string;
  completed_at?: string;
}

export interface DocumentMetadata {
  student_id?: string;
  student_name?: string;
  document_date?: string;
  document_number?: string;
  extra?: {
    class_name?: string;
    faculty?: string;
    document_type?: string;
    reason?: string;
    [key: string]: any;
  };
}

export interface DocumentDetail extends DocumentSummary {
  page_count?: number;
  category_id?: string;
  uploaded_by: string;
  is_deleted: boolean;
  updated_at: string;
  ocr_result?: OCRResult;
  processing_job?: ProcessingJob;
  metadata?: DocumentMetadata;
}

export interface VerificationData {
  is_valid: boolean;
  document_id: string;
  title: string;
  original_filename: string;
  ocr_status: string;
  student_name?: string;
  student_id?: string;
  approved_at?: string;
  approved_by_name?: string;
  verification_code: string;
  qr_payload: string;
  issued_by: string;
}

export interface UploadResponse {
  id: string;
  title: string;
  original_filename: string;
  file_type: string;
  file_size_bytes: number;
  ocr_status: string;
  minio_object_key: string;
  job_id?: string;
  celery_task_id?: string;
  message: string;
  created_at: string;
}

export interface ListDocumentsParams {
  page?: number;
  pageSize?: number;
  ocrStatus?: string;
  categoryId?: string;
  search?: string;
}

export const documentsApi = {
  list: async (params?: ListDocumentsParams): Promise<DocumentListResponse> => {
    const res = await apiClient.get<DocumentListResponse>("/documents", {
      params: {
        page: params?.page || 1,
        page_size: params?.pageSize || 20,
        ocr_status: params?.ocrStatus,
        category_id: params?.categoryId,
        search: params?.search,
      },
    });
    return res.data;
  },

  upload: async (file: File, title?: string, categoryId?: string): Promise<UploadResponse> => {
    const formData = new FormData();
    formData.append("file", file);
    if (title) formData.append("title", title);
    if (categoryId) formData.append("category_id", categoryId);

    const res = await apiClient.post<UploadResponse>("/documents/upload", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });
    return res.data;
  },

  getById: async (id: string): Promise<DocumentDetail> => {
    const res = await apiClient.get<DocumentDetail>(`/documents/${id}`);
    return res.data;
  },

  saveCorrection: async (id: string, correctedText: string): Promise<DocumentDetail> => {
    const res = await apiClient.put<DocumentDetail>(`/documents/${id}/ocr-correction`, {
      corrected_text: correctedText,
    });
    return res.data;
  },

  reprocessOCR: async (id: string): Promise<{ message: string; document_id: string; status: string }> => {
    const res = await apiClient.post(`/documents/${id}/reprocess-ocr`);
    return res.data;
  },

  triggerExtractFields: async (id: string): Promise<DocumentMetadata> => {
    const res = await apiClient.post<DocumentMetadata>(`/documents/${id}/extract-fields`);
    return res.data;
  },

  getVerification: async (id: string): Promise<VerificationData> => {
    const res = await apiClient.get<VerificationData>(`/documents/${id}/verification`);
    return res.data;
  },

  getPublicVerification: async (id: string): Promise<VerificationData> => {
    const res = await apiClient.get<VerificationData>(`/verify/${id}`);
    return res.data;
  },

  approve: async (id: string): Promise<DocumentDetail> => {
    const res = await apiClient.patch<DocumentDetail>(`/documents/${id}/approve`);
    return res.data;
  },

  reject: async (id: string): Promise<DocumentDetail> => {
    const res = await apiClient.patch<DocumentDetail>(`/documents/${id}/reject`);
    return res.data;
  },

  exportExcelUrl: (status?: string, categoryId?: string) => {
    const params = new URLSearchParams();
    if (status && status !== "ALL") params.append("ocr_status", status);
    if (categoryId) params.append("category_id", categoryId);
    return `${API_BASE_URL}/documents/export/excel?${params.toString()}`;
  },
};

File: backend/app/schemas/documents.py
"""
backend/app/schemas/documents.py — Pydantic Schemas for Documents

Quy ước đặt tên theo .ai/CODING_RULES.md:
- <Entity>Create
- <Entity>Update
- <Entity>Response
- <Entity>DetailResponse
"""
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ── Sub-schemas ──────────────────────────────────────────────────────────────
class OCRResultResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    document_id: UUID
    is_latest: bool
    raw_text: str | None = None
    corrected_text: str | None = None
    confidence_score: float | None = None
    ocr_engine: str
    ocr_engine_version: str | None = None
    page_texts: list[dict] | dict | None = None
    processing_time_ms: int | None = None
    is_corrected: bool = False
    corrected_by: UUID | None = None
    corrected_at: datetime | None = None
    created_at: datetime


class ProcessingJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    document_id: UUID
    celery_task_id: str | None = None
    status: str
    error_message: str | None = None
    retry_count: int
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime


class DocumentMetadataResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    document_id: UUID
    student_id: str | None = None
    student_name: str | None = None
    document_date: datetime | None = None
    document_number: str | None = None
    extra: dict | None = None


class OCRCorrectionRequest(BaseModel):
    """Payload gửi lên khi cán bộ hiệu chỉnh văn bản OCR."""
    corrected_text: str = Field(..., min_length=1, description="Nội dung văn bản sau khi hiệu chỉnh")


class FieldExtractionUpdateRequest(BaseModel):
    """Payload cập nhật metadata bóc tách thông tin sinh viên."""
    student_id: str | None = None
    student_name: str | None = None
    document_number: str | None = None
    document_date: datetime | None = None
    extra: dict | None = None


class VerificationResponse(BaseModel):
    """Payload phản hồi khi quét mã QR xác thực hồ sơ."""
    is_valid: bool
    document_id: UUID
    title: str
    original_filename: str
    ocr_status: str
    student_name: str | None = None
    student_id: str | None = None
    approved_at: datetime | None = None
    approved_by_name: str | None = None
    verification_code: str
    qr_payload: str
    issued_by: str = "Trường Đại học Đà Lạt - Phòng Công tác Sinh viên (DocuCTSV)"


# ── Main Document Schemas ────────────────────────────────────────────────────
class DocumentUploadResponse(BaseModel):
    """Response trả về ngay sau khi upload (202 Accepted)."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    original_filename: str
    file_type: str
    file_size_bytes: int
    ocr_status: str
    minio_object_key: str
    job_id: UUID | None = None
    celery_task_id: str | None = None
    message: str = "Tài liệu đã được tải lên và đưa vào hàng đợi OCR."
    created_at: datetime


class DocumentResponse(BaseModel):
    """Thông tin cơ bản của Document."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    original_filename: str
    file_type: str
    file_size_bytes: int
    page_count: int | None = None
    category_id: UUID | None = None
    uploaded_by: UUID | None = None
    ocr_status: str
    minio_object_key: str | None = None
    is_deleted: bool = False
    created_at: datetime
    updated_at: datetime


class DocumentListItem(DocumentResponse):
    """Thông tin item trong danh sách tài liệu."""
    uploader_name: str | None = None
    uploader_mssv: str | None = None
    category_name: str | None = None
    category_code: str | None = None
    confidence_score: float | None = None
    ocr_confidence: float | None = None
    student_id: str | None = None
    student_name: str | None = None


class DocumentListResponse(BaseModel):
    """Danh sách tài liệu có phân trang."""
    items: list[DocumentListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


class DocumentDetailResponse(DocumentResponse):
    """Thông tin chi tiết tài liệu kèm kết quả OCR mới nhất và trạng thái Job."""
    ocr_result: OCRResultResponse | None = None
    processing_job: ProcessingJobResponse | None = None
    metadata_: DocumentMetadataResponse | None = Field(default=None, alias="metadata")
